Add Expert Parallelism (EP) Testcases support to MoE testing framework

amolkerkar opened the pull request:
# MoE Testing Framework Updates

## High-Level Summary

Enhanced the MoE (Mixture of Experts) testing framework in `utils_neuron.py` to support comprehensive Expert Parallelism (EP) testing alongside existing Tensor Parallelism (TP) configurations. The changes enable testing of various parallelism strategies for large-scale MoE models including 150B and Qwen3-235B configurations.

## Key Changes

### 1. Enhanced `build_name` Function
- **Purpose**: Improved test case naming to distinguish between TP and EP configurations
- **Changes**: 
  - Added support for `expert` mesh specifications in naming
  - Enhanced mesh string generation to show `ep{N}` for expert parallelism vs `tp{N}` for tensor parallelism
  - Better identification of test configurations in logs and results

### 2. Updated `build_grid_space_input_hidden` Function
- **Purpose**: Core function for generating test configurations with various parallelism combinations
- **Key Updates**:
  - **TP/EP Combinations**: Supports combinations like (1,16), (1,64), (4,1), (4,16), (16,1), (64,1)
  - **Constraint**: Maximum product of TP×EP ≤ 64 (Trn2 architecture limit)
  - **Mesh Specification Logic**:
    - EP > 1: `{"fsdp": -1, "expert": ep_degree}` with `n_groups = ep_degree`
    - TP only: `{"fsdp": -1, "model": tp_degree}` with `n_groups ∈ [1, 4]`
  - **Batch Size Mapping**: Appropriate batch sizes for different parallelism degrees (64→1, 16→4, 4→16)

### 3. Enhanced `build_grid_space_150B` Function
- **Purpose**: Specific test configurations for 150B model scale
- **New Features**:
  - **EP Test Cases**: Added EP=16 configurations for 16-expert scenarios
  - **Configurations**:
    - EP=16 with TP=1: `mesh_spec={"fsdp": -1, "expert": 16}`
    - EP=16 with TP=4: `mesh_spec={"fsdp": -1, "expert": 16, "model": 4}`
  - **Model Dimensions**: 6144→15360 hidden dimensions
  - **Expert Counts**: Supports 8 and 16 experts with appropriate group configurations

### 4. New `build_grid_space_qwen3_235b` Function
- **Purpose**: Comprehensive testing for Qwen3-235B model with both TP and EP strategies
- **Configurations**:
  - **TP=64**: `mesh_spec={"fsdp": -1, "model": 64}`, `n_groups=1`, `batch=1`
  - **EP=16**: `mesh_spec={"fsdp": -1, "expert": 16}`, `n_groups=16`, `batch=4`
  - **EP=64**: `mesh_spec={"fsdp": -1, "expert": 64}`, `n_groups=64`, `batch=1`
- **Model Specifications**:
  - Input dimension: 4096
  - Hidden dimension: 12288
  - Experts: 128
  - Sequence length: 16384
  - Top-k values: [1, 8]

## Technical Details

### Parallelism Strategy Logic
- **Expert Parallelism (EP)**: `n_groups = ep_degree` ensures proper load balancing across expert-parallel processes
- **Tensor Parallelism (TP)**: `n_groups ∈ [1, 4]` for different input splitting strategies
- **Constraint Enforcement**: `n_experts ≥ ep_degree` to ensure sufficient experts per EP process

### Batch Size Optimization
```python
batch_sizes = {
    4: 16,   # Lower parallelism → higher batch size
    8: 8,
    16: 4,   # EP=16 uses batch=4
    32: 2,
    64: 1,   # EP=64 uses batch=1 (highest parallelism)
}
```

### Test Coverage
- **Switch-Base**: Uses `build_grid_space_input_hidden` with 1536→6144 dimensions
- **150B Models**: Uses `build_grid_space_150B` with 6144→15360 dimensions  
- **Qwen3-235B**: Uses `build_grid_space_qwen3_235b` with 4096→12288 dimensions

## Usage

### Running Tests
```bash
# Switch-base with EP configurations
sbatch aws-testing/test.slurm switch-base

# 150B development tests
sbatch aws-testing/test.slurm 150b

# Qwen3-235B with TP/EP combinations
sbatch aws-testing/test.slurm qwen3-235b
```

### Expected Test Cases
- **Switch-Base**: ~32 test configurations covering EP=16, TP=4, TP=16 with various expert counts
- **150B**: Enhanced with EP=16 test cases for 16-expert scenarios
- **Qwen3-235B**: 6 test configurations (3 parallelism strategies × 2 top-k values)

amolkerkar committed:
Add Expert Parallelism (EP) Testcases support to MoE testing framework

rahul003 commented on aws-testing/utils_neuron.py:
```diff
@@ -494,31 +500,57 @@ def build_grid_space_input_hidden(self, input_dim=2048, hidden_dim=7168, min_seq
+        
+        for tp_degree, ep_degree in tp_ep_combinations:
+            if min_tp is not None and tp_degree < min_tp:
+                continue
```
respect min_tp only when ep_degree is 1 for now, we can filter more as needed. min_tp existed because it wouldn't fit the model; but with EP now it can

rahul003 commented on aws-testing/utils_neuron.py:
```diff
@@ -494,31 +500,57 @@ def build_grid_space_input_hidden(self, input_dim=2048, hidden_dim=7168, min_seq
+            #mesh_spec and batch based on parallelism type
+            if ep_degree > 1:
+                mesh_spec = {"fsdp": -1, "expert": ep_degree}
+                batch = batch_sizes[ep_degree]
```
batch size I think should still come based on tp_degree

rahul003 commented on aws-testing/utils_neuron.py:
```diff
@@ -582,11 +614,19 @@ def build_grid_space_qwen3_235b(self):
+        
+        mesh_configs = [
+            ({"fsdp": -1, "model": 64}, 1, 1),      # TP=64, batch=1, n_groups=1
+            ({"fsdp": -1, "expert": 16}, 4, 16),    # EP=16, batch=4, n_groups=16  
```
for any config with tp=4, use bs=16; tp=16, bs=4; tp=64, bs=1

amolkerkar replied:
is this a valid config for build_grid_space_input_hidden:
```
        mesh_configs = [
            # TP-only configurations
            ({"fsdp": -1, "model": 4}, 16, 1),      # TP=4, batch=16, n_groups=1
            ({"fsdp": -1, "model": 16}, 4, 1),      # TP=16, batch=4, n_groups=1
            ({"fsdp": -1, "model": 64}, 1, 1),      # TP=64, batch=1, n_groups=1
            
            # EP-only configurations
            ({"fsdp": -1, "expert": 16}, 16, 16),   # EP=16, batch=16, n_groups=16
            ({"fsdp": -1, "expert": 64}, 16, 64),   # EP=64, batch=16, n_groups=64
            
            # Mixed TP+EP configurations
            ({"fsdp": -1, "model": 4, "expert": 16}, 16, 16),  # TP=4×EP=16, batch=16, n_groups=16
        ]
```

rahul003 commented on aws-testing/utils_neuron.py:
```diff
@@ -582,11 +614,19 @@ def build_grid_space_qwen3_235b(self):
-        # TODO add EP
-        for mesh_spec in [{"fsdp": -1, "model": 64}]:
-            batch = 4 if mesh_spec["model"] == 16 else 1
+        
```
also add EP tests for the other build_grid_space functions like switch-xxl, llama, etc

amolkerkar committed:
Add comprehensive EP support to MoE testing functions

- Updated build_grid_space_qwen3_235b with all 6 valid TP×EP combinations
- Updated build_grid_space_llama4_maverick with all 6 valid TP×EP combinations
- Updated build_grid_space_switch_xxl with all 6 valid TP×EP combinations
- All functions now test TP-only, EP-only, and mixed TP+EP configurations
- Follows TP×EP ≤ 64 constraint and proper batch size/n_groups rules

amolkerkar committed:
Add EP configurations to switch-base test suite

- Modified build_grid_space_input_hidden() to include EP=16 and EP=64 configurations
- Added (tp,ep) combinations: (1,16), (1,64), (4,16), (4,1), (16,1), (64,1)
- Set n_groups=ep_degree for EP configs, capacity_factor=2 for all configs
- Fixed build_name() function to handle both TP and EP mesh specifications
- Fixed syntax errors in build_grid_space_150B function

amolkerkar committed:
Fix build_name for mixed TP+EP and add hybrid EP support

- Fixed build_name() to properly handle mixed TP+EP mesh specs (tp4ep16 format)
- Added hybrid EP support to build_grid_space_12B() with 4 EP tests
- Added hybrid EP support to build_grid_space_50B() with 8 EP tests
- Added EP support to build_grid_space_150B() with 5 EP tests
- All functions preserve existing TP test coverage while adding EP functionality
- Mixed TP+EP configurations now generate unique test names

rahul003 reviewed (approved):
Lets merge
